applications/nlp/transformer/train.py

```diff
@@ -99,53 +86,25 @@ def make_model(
         name="prediction_layer"
     )
     preds = lbann.ChannelwiseSoftmax(preds)
-    preds = lbann.Slice(preds, axis=0, slice_points=range(sequence_length))
-    preds = [lbann.Identity(preds) for _ in range(sequence_length-1)]
+    preds = lbann.TensorPermute(preds, axes=[1, 0])
 
-    # Count number of non-pad tokens
+    # Compute labels
     label_tokens = lbann.Identity(lbann.Slice(
         input_,
         slice_points=[sequence_length+1, 2*sequence_length],
     ))
-    pads = lbann.Constant(value=pad_index, num_neurons=sequence_length-1)
-    is_not_pad = lbann.NotEqual(label_tokens, pads)
-    num_not_pad = lbann.Reduction(is_not_pad, mode='sum')
-
-    # Cross entropy loss with label smoothing
-    label_tokens = lbann.Slice(
-        label_tokens,
-        slice_points=range(sequence_length),
-    )
-    label_tokens = [lbann.Identity(label_tokens) for _ in range(sequence_length-1)]
-    if label_smoothing > 0:
-        uniform_label = lbann.Constant(
-            value=1/vocab_size,
-            num_neurons=[1, vocab_size]
-        )
-    loss = []
-    for i in range(sequence_length-1):
-        label = lbann.OneHot(label_tokens[i], size=vocab_size)
-        label = lbann.Reshape(label, dims=[1, vocab_size])
-        if label_smoothing > 0:
-            label = lbann.WeightedSum(
-                label,
-                uniform_label,
-                scaling_factors=[1-label_smoothing, label_smoothing],
-            )
-        loss.append(lbann.CrossEntropy(preds[i], label))
-    loss = lbann.Concatenation(loss)
+    labels = lbann.Reshape(label_tokens, dims=[1, sequence_length - 1])
 
-    # Average cross entropy over non-pad tokens
-    loss_scales = lbann.Divide(
-        is_not_pad,
-        lbann.Tessellate(num_not_pad, hint_layer=is_not_pad),
-    )
-    loss = lbann.Multiply(loss, loss_scales)
-    loss = lbann.Reduction(loss, mode='sum')
+    # Filter out output predictions that are in padding from cross-entropy by
+    # using values that will never contribute to the cross-entropy loss
+    labels = lbann.Select(labels, lbann.Identity(labels), value=pad_index, if_false=(vocab_size + 1))
+
+    # Compute cross-entropy
+    loss = lbann.CrossEntropy(preds, labels, use_labels=True)
 
     # Construct model
     metrics = []
-    callbacks = [lbann.CallbackPrint(), lbann.CallbackTimer()]
+    callbacks = [lbann.CallbackPrint(), lbann.CallbackTimer(), lbann.CallbackProfiler()]
```

Comment:
we might want to remove this change

Reply:
I think that we can leave it in for now.